webnn/resources/utils.js

```diff
@@ -93,52 +35,288 @@ function loadTestData(file) {
-}
+// Refer to precision metrics on https://github.com/webmachinelearning/webnn/issues/265#issuecomment-1256242643
+const PrecisionMetrics = {
+  batchNormalization: {ULP: {float32: 6, float16: 6}},
+  clamp: {ULP: {float32: 0, float16: 0}},
+  concat: {ULP: {float32: 0, float16: 0}},
+  // conv2d IEPOE*2 ULP
+  // element-wise binary operations
+  add: {ULP: {float32: 1, float16: 1}},
+  sub: {ULP: {float32: 1, float16: 1}},
+  mul: {ULP: {float32: 1, float16: 1}},
+  div: {ULP: {float32: 2, float16: 2}},
+  max: {ULP: {float32: 0, float16: 0}},
+  min: {ULP: {float32: 0, float16: 0}},
+  pow: {ULP: {float32: 32, float16: 2}},
+  // element-wise unary operations
+  abs: {ULP: {float32: 0, float16: 0}},
+  ceil: {ULP: {float32: 0, float16: 0}},
+  cos: {ATOL: {float32: 1/1024, float16: 1/512}},
+  exp: {ULP: {float32: 32, float16: 1}},
+  floor: {ULP: {float32: 0, float16: 0}},
+  log: {ATOL: {float32: 1/1024, float16: 1/1024}},
+  neg: {ULP: {float32: 0, float16: 0}},
+  sin: {ATOL: {float32: 1/1024, float16: 1/512}},
+  tan: {ATOL: {float32: 1/1024, float16: 1/512}},
+  // gemm IEPOE*2+3 ULP
+  leakRelu: {ULP: {float32: 1, float16: 1}},
+  // matmul IEPOE*2 ULP
+  // pooling operations
+  // averagePool2d IEPOE+2 ULP
+  maxPool2d: {ULP: {float32: 0, float16: 0}},
+  // reduction operations
+  reduceMax: {ULP: {float32: 0, float16: 0}},
+  // reduceMean IEPOE+2 ULP
+  reduceMin: {ULP: {float32: 0, float16: 0}},
+  // reduceProduct IEPOE ULP
+  // reduceSum IEPOE ULP
+  relu: {ULP: {float32: 0, float16: 0}},
+  reshape: {ULP: {float32: 0, float16: 0}},
+  // float32 (leaving a few ULP for roundoff)
+  sigmoid: {ULP: {float32: 32+2, float16: 3}},
+  slice: {ULP: {float32: 0, float16: 0}},
+  // softmax IEPOE*2+3 ULP
+  split: {ULP: {float32: 0, float16: 0}},
+  squeeze: {ULP: {float32: 0, float16: 0}},
+  tanh: {ATOL: {float32: 1/1024, float16: 1/512}},
+  transpose: {ULP: {float32: 0, float16: 0}},
+};
 
 /**
- * Asynchronously create context and builder.
- * @param {MLContextOptions} contextOptions
-* @returns {Promise([MLContext, MLGraphBuilder])}
+ * Run WebNN operation tests.
+ * @param {String} opName - An operation name
+ * @param {Array} tests - A test resources array
+ * @param {Object} buildGraphFunc - A function for building a graph
  */
-async function createContextAndBuilder(contextOptions) {
-  context = await navigator.ml.createContext(contextOptions);
-  builder = new MLGraphBuilder(context);
-  return [context, builder];
-}
+function testWebNNOperation(opName, tests, buildGraphFunc) {
+  /**
+   * Get ULP tolerances of gemm operation.
+   * @param {Array} shapeA [number, number]
+   * @param {Object} options
+   * @param {MLOperand | Number} [options.c]
+   * @param {Number} [options.alpha]
+   * @param {Number} [options.beta]
+   * @param {Boolean} [options.aTranspose]
+   * @param {Boolean} [options.bTranspose]
+   * @returns {Number} A tolerance number
+   */
+  function getGemmPrecisionTolerance(shapeA, options) {
```

Comment:
Is there any reason that we have to put these functions inside the `testWebNNOperation`? This would make code not easy to read.

Reply:
Thanks for your suggestion, moved these functions outside.